webapp-django/crashstats/signature/views.py

```diff
@@ -424,3 +424,112 @@ def signature_graph_data(request, channel):
+        # There was an error in the form, let's return it.
+        return params
+
+    signature = params['signature'][0]
+
+    data = {}
+
+    current_query = request.GET.copy()
+    if 'page' in current_query:
+        del current_query['page']
+
+    data['params'] = current_query.copy()
+
+    params['signature'] = '=' + signature
+    params['_results_number'] = 0
+    params['_facets'] = [
+        'platform',
+        'cpu_name',
+        'process_type',
+        'flash_version',
+    ]
+    params['_histogram.uptime'] = ['product']
+    params['_histogram_interval.uptime'] = 60
+
+    # If the user has permissions, show exploitability.
+    all_fields = SuperSearchFields().get()
+    if has_permissions(
+        request.user, all_fields['exploitability']['permissions_needed']
+    ):
+        params['_histogram.date'] = ['exploitability']
+
+    api = SuperSearchUnredacted()
+    try:
+        search_results = api.get(**params)
+    except models.BadStatusCodeError, e:
+        # We need to return the error message in some HTML form for jQuery to
+        # pick it up.
+        return http.HttpResponseBadRequest('<ul><li>%s</li></ul>' % e)
+
+    facets = search_results['facets']
+
+    # Transform uptime data to be easier to consume.
+    # Keys are in minutes.
+    if 'histogram_uptime' in facets:
+        uptimes_count = {
+            0: 0,
+            1: 0,
+            5: 0,
+            15: 0,
+            60: 0,
+        }
+        for uptime in facets['histogram_uptime']:
+            for uptime_minutes in sorted(uptimes_count.keys(), reverse=True):
+                uptime_seconds = uptime_minutes * 60
+
+                if uptime['term'] >= uptime_seconds:
+                    uptimes_count[uptime_minutes] += uptime['count']
+                    break
+
+        labels = {
+            0: '< 1 min',
+            1: '1-5 min',
+            5: '5-15 min',
+            15: '15-60 min',
+            60: '> 1 hour'
+        }
+        uptimes = [
+            {'term': labels.get(key), 'count': count}
+            for key, count in uptimes_count.items()
+            if count > 0
+        ]
+        uptimes = sorted(uptimes, key=lambda x: x['count'], reverse=True)
+        data['uptimes'] = uptimes
+
+        # That's a big piece of data we don't need to pass to the template.
+        del facets['histogram_uptime']
+
+    # Transform exploitability facet.
+    if 'histogram_date' in facets:
+        exploitability_base = {
+            'none': 0,
+            'low': 0,
+            'medium': 0,
+            'high': 0,
+        }
+        for day in facets['histogram_date']:
+            exploitability = dict(exploitability_base)
+            for expl in day['facets']['exploitability']:
+                if expl['term'] in exploitability:
+                    exploitability[expl['term']] = expl['count']
+            day['exploitability'] = exploitability
+            del day['facets']['exploitability']
```

Comment:
Why delete? That memory will get garbage collected at the end of the wsgi request.